File: iEngine-Web/src/views/controls/FirstPersonControls.ts
import { PerspectiveCamera } from '../cameras/PerspectiveCamera';
import { BaseControls } from './BaseControls';

export class FirstPersonControls extends BaseControls {
    private camera: PerspectiveCamera;
    private domElement: HTMLElement;
    private isDragging = false;
    private lastX = 0;
    private lastY = 0;
    private yaw = 0;
    private pitch = 0;
    private moveSpeed = 0.1;
    private lookSpeed = 0.002;

    constructor(camera: PerspectiveCamera, domElement: HTMLElement) {
        super();

        //
        this.camera = camera;
        this.domElement = domElement;

        // 初始化 yaw/pitch 与当前相机方向同步
        const dir = [
            camera.target.x - camera.position.x,
            camera.target.y - camera.position.y,
            camera.target.z - camera.position.z
        ];
        const len = Math.sqrt(dir[0] * dir[0] + dir[1] * dir[1] + dir[2] * dir[2]);
        if (len > 0) {
            this.yaw = Math.atan2(dir[0], dir[2]);
            this.pitch = Math.asin(dir[1] / len);
        }

        // 保证初始状态
        this.updateCamera();

        domElement.addEventListener('mousedown', this.onMouseDown, { passive: false });
        domElement.addEventListener('wheel', this.onMouseWheel, { passive: false });
        window.addEventListener('keydown', this.onKeyDown, { passive: false });
    }

    private onMouseDown = (e: MouseEvent) => {
        this.isDragging = true;
        this.lastX = e.clientX;
        this.lastY = e.clientY;
        window.addEventListener('mousemove', this.onMouseMove, { passive: false });
        window.addEventListener('mouseup', this.onMouseUp, { passive: false });
    };

    private onMouseMove = (e: MouseEvent) => {
        if (!this.isDragging) return;
        const dx = e.clientX - this.lastX;
        const dy = e.clientY - this.lastY;
        this.lastX = e.clientX;
        this.lastY = e.clientY;

        this.yaw -= dx * this.lookSpeed;
        this.pitch -= dy * this.lookSpeed;
        this.pitch = Math.max(-Math.PI / 2, Math.min(Math.PI / 2, this.pitch));
        this.updateCamera();
    };

    private onMouseUp = () => {
        this.isDragging = false;
        window.removeEventListener('mousemove', this.onMouseMove);
        window.removeEventListener('mouseup', this.onMouseUp);
    };

    private onMouseWheel = (e: WheelEvent) => {
        // 可用于前后移动
        this.camera.position.z += e.deltaY * 0.01;
    };

    private onKeyDown = (e: KeyboardEvent) => {
        // 简单的WASD移动
        const forward = [
            Math.sin(this.yaw),
            0,
            Math.cos(this.yaw)
        ];
        const right = [
            Math.cos(this.yaw),
            0,
            -Math.sin(this.yaw)
        ];
        if (e.key === 'w') {
            this.camera.position.x += forward[0] * this.moveSpeed;
            this.camera.position.z += forward[2] * this.moveSpeed;
        }
        if (e.key === 's') {
            this.camera.position.x -= forward[0] * this.moveSpeed;
            this.camera.position.z -= forward[2] * this.moveSpeed;
        }
        if (e.key === 'a') {
            this.camera.position.x -= right[0] * this.moveSpeed;
            this.camera.position.z -= right[2] * this.moveSpeed;
        }
        if (e.key === 'd') {
            this.camera.position.x += right[0] * this.moveSpeed;
            this.camera.position.z += right[2] * this.moveSpeed;
        }
        this.updateCamera();
    };

    private updateCamera() {
        // 计算朝向
        const dir = [
            Math.sin(this.yaw) * Math.cos(this.pitch),
            Math.sin(this.pitch),
            Math.cos(this.yaw) * Math.cos(this.pitch)
        ];
        const lookAt = [
            this.camera.position.x  + dir[0],
            this.camera.position.y  + dir[1],
            this.camera.position.z  + dir[2]
        ];
        this.camera.lookAt(lookAt[0], lookAt[1], lookAt[2]);
    }

    dispose() {
        this.domElement.removeEventListener('mousedown', this.onMouseDown);
        this.domElement.removeEventListener('wheel', this.onMouseWheel);
        window.removeEventListener('mousemove', this.onMouseMove);
        window.removeEventListener('mouseup', this.onMouseUp);
        window.removeEventListener('keydown', this.onKeyDown);
    }
}

File: iEngine-Native/Engine/src/views/controls/FirstPersonControls.cpp
#include "iengine/views/controls/FirstPersonControls.h"
#include "../../../include/iengine/views/controls/FirstPersonControls.h"
#include "iengine/windowing/Window.h"
#include <algorithm>
#include <cmath>
#include <iostream>

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

namespace iengine {
    FirstPersonControls::FirstPersonControls(std::shared_ptr<PerspectiveCamera> camera, std::shared_ptr<WindowInterface> window)
        : camera_(camera), window_(window) {
        
        // 初始化 yaw/pitch 让相机朝向原点 (0,0,0)
        // 从位置 (0,0,5) 朝向 (0,0,0)，需要 yaw=π（180度转身）
        yaw_ = static_cast<float>(M_PI);  // 180度，朝向 -Z 方向
        pitch_ = 0.0f;  // 水平朝向
        
        // 从相机获取当前位置，而不是硬编码
        // TODO: 由于position_是protected，暂时使用默认值，但需要与场景设置保持一致
        currentPosition_ = Vector3(0.0f, 0.0f, 5.0f);
        camera_->setPosition(currentPosition_.x, currentPosition_.y, currentPosition_.z);
        
        std::cout << "FirstPersonControls: 相机位置设置为 (" << currentPosition_.x 
                  << ", " << currentPosition_.y << ", " << currentPosition_.z << ")" << std::endl;
        
        // 使用lambda捕获this指针来绑定事件处理器
        window_->setEventCallback([this](const WindowEvent& event) {
            this->handleWindowEvent(event);
        });
        
        std::cout << "FirstPersonControls: 初始化完成，支持WASD移动和鼠标视角控制" << std::endl;
        
        // 保证初始状态
        updateCamera();
    }
    
    FirstPersonControls::~FirstPersonControls() {
        dispose();
    }
    
    void FirstPersonControls::dispose() {
        // 移除事件监听，重置窗口事件回调
        if (window_) {
            window_->setEventCallback(nullptr);
        }
    }
    
    bool FirstPersonControls::onWindowEvent(const WindowEvent& event) {
        // 直接调用现有的处理逻辑
        handleWindowEvent(event);
        
        // 返回false表示不阻断事件传播，允许其他监听器处理
        return false;
    }
    
    void FirstPersonControls::handleWindowEvent(const WindowEvent& event) {
        switch (event.type) {
            case WindowEventType::MouseButton:
                std::cout << "FirstPersonControls: 收到鼠标按键事件 - 按键:" << (int)event.data.mouseButton.button 
                         << ", 动作:" << (int)event.data.mouseButton.action 
                         << ", 位置:(" << event.data.mouseButton.x << ", " << event.data.mouseButton.y << ")" << std::endl;
                if (event.data.mouseButton.button == MouseButton::Left) {
                    if (event.data.mouseButton.action == KeyAction::Press) {
                        onMouseDown(static_cast<float>(event.data.mouseButton.x), 
                                   static_cast<float>(event.data.mouseButton.y));
                    } else if (event.data.mouseButton.action == KeyAction::Release) {
                        onMouseUp();
                    }
                }
                break;
            case WindowEventType::MouseMove:
                if (isDragging_) {
                    std::cout << "FirstPersonControls: 鼠标拖拽中 - 位置:(" << event.data.mouseMove.x << ", " << event.data.mouseMove.y << ")" << std::endl;
                }
                onMouseMove(static_cast<float>(event.data.mouseMove.x), 
                           static_cast<float>(event.data.mouseMove.y));
                break;
            case WindowEventType::MouseScroll: {
                std::cout << "FirstPersonControls: 收到滚轮事件 - 偏移:" << event.data.mouseScroll.yoffset << std::endl;
                // 防御性判断：限制滚轮值范围以防止异常值
                float wheelDelta = static_cast<float>(event.data.mouseScroll.yoffset);
                if (std::abs(wheelDelta) > 10.0f) {
                    wheelDelta = wheelDelta > 0 ? 10.0f : -10.0f;
                }
                // 用于前后移动
                onMouseWheel(wheelDelta);
                break;
            }
            case WindowEventType::Key:
                std::cout << "FirstPersonControls: 收到键盘事件 - 按键:" << event.data.key.key 
                         << ", 动作:" << (int)event.data.key.action << std::endl;
                onKeyDown(event.data.key.key, (int)event.data.key.action);
                break;
            default:
                // 其他事件不处理
                break;
        }
    }
    
    void FirstPersonControls::onMouseDown(float x, float y) {
        isDragging_ = true;
        lastX_ = x;
        lastY_ = y;
        std::cout << "FirstPersonControls: 开始拖拽 - 位置:(" << x << ", " << y << ")" << std::endl;
    }
    
    void FirstPersonControls::onMouseMove(float x, float y) {
        if (!isDragging_) return;
        const float dx = x - lastX_;
        const float dy = y - lastY_;
        lastX_ = x;
        lastY_ = y;
        
        yaw_ -= dx * lookSpeed_;
        pitch_ -= dy * lookSpeed_;
        pitch_ = std::max(-static_cast<float>(M_PI) / 2, std::min(static_cast<float>(M_PI) / 2, pitch_));
        updateCamera();
    }
    
    void FirstPersonControls::onMouseUp() {
        isDragging_ = false;
        std::cout << "FirstPersonControls: 停止拖拽" << std::endl;
    }
    
    void FirstPersonControls::onMouseWheel(float delta) {
        // 防御性判断：如果值过小则忽略，防止微小抖动
        if (std::abs(delta) < 0.001f) {
            return;
        }
        
        // 用于前后移动（参考TS版本）
        // 计算前方向
        const float forwardX = std::sin(yaw_);
        const float forwardZ = std::cos(yaw_);
        
        // 限制移动速度
        float moveAmount = delta * 0.01f;
        moveAmount = std::max(-1.0f, std::min(1.0f, moveAmount));
        
        // 获取当前相机位置并移动（直接使用setPosition方法）
        // 需要先获取当前位置，但由于position_是protected，我们维护本地位置状态
        currentPosition_.x += forwardX * moveAmount;
        currentPosition_.z += forwardZ * moveAmount;
        camera_->setPosition(currentPosition_.x, currentPosition_.y, currentPosition_.z);
        updateCamera();
    }
    
    void FirstPersonControls::onKeyDown(int key, int action) {
        // 只在按下时处理（不处理重复事件）
        if (action != (int)KeyAction::Press) {
            return;
        }
        
        // 简单的WASD移动（参考TS版本）
        const float forwardX = std::sin(yaw_);
        const float forwardZ = std::cos(yaw_);
        const float rightX = std::cos(yaw_);
        const float rightZ = -std::sin(yaw_);
        
        // GLFW键码对应（参考GLFW定义）
        const int GLFW_KEY_W = 87;
        const int GLFW_KEY_S = 83;
        const int GLFW_KEY_A = 65;
        const int GLFW_KEY_D = 68;
        
        if (key == GLFW_KEY_W) {
            std::cout << "FirstPersonControls: W键 - 向前移动" << std::endl;
            currentPosition_.x += forwardX * moveSpeed_;
            currentPosition_.z += forwardZ * moveSpeed_;
            camera_->setPosition(currentPosition_.x, currentPosition_.y, currentPosition_.z);
        }
        if (key == GLFW_KEY_S) {
            std::cout << "FirstPersonControls: S键 - 向后移动" << std::endl;
            currentPosition_.x -= forwardX * moveSpeed_;
            currentPosition_.z -= forwardZ * moveSpeed_;
            camera_->setPosition(currentPosition_.x, currentPosition_.y, currentPosition_.z);
        }
        if (key == GLFW_KEY_A) {
            std::cout << "FirstPersonControls: A键 - 向左移动" << std::endl;
            currentPosition_.x -= rightX * moveSpeed_;
            currentPosition_.z -= rightZ * moveSpeed_;
            camera_->setPosition(currentPosition_.x, currentPosition_.y, currentPosition_.z);
        }
        if (key == GLFW_KEY_D) {
            std::cout << "FirstPersonControls: D键 - 向右移动" << std::endl;
            currentPosition_.x += rightX * moveSpeed_;
            currentPosition_.z += rightZ * moveSpeed_;
            camera_->setPosition(currentPosition_.x, currentPosition_.y, currentPosition_.z);
        }
        updateCamera();
    }
    
    void FirstPersonControls::updateCamera() {
        // 计算朝向（参考TS版本）
        const float dirX = std::sin(yaw_) * std::cos(pitch_);
        const float dirY = std::sin(pitch_);
        const float dirZ = std::cos(yaw_) * std::cos(pitch_);
        
        // 计算目标点
        const float lookAtX = currentPosition_.x + dirX;
        const float lookAtY = currentPosition_.y + dirY;
        const float lookAtZ = currentPosition_.z + dirZ;
        
        // 设置相机朝向
        camera_->lookAt(lookAtX, lookAtY, lookAtZ);
    }
}